src/bitset.c

```diff
@@ -450,7 +450,73 @@ bool bitset_trim(bitset_t *bitset) {
     return true;
 }
 
+/* iterate over the set bits
+ like so :
+   size_t buffer[256];
+   size_t howmany = 0;
+  for(size_t startfrom = 0; (howmany = bitset_next_set_bits(b,buffer,256, &startfrom)) >
+ 0 ; startfrom++) {
+    //.....
+  }
+  */
+size_t bitset_next_set_bits(const bitset_t *bitset, size_t *buffer,
+                            size_t capacity, size_t *startfrom) {
+    if (capacity == 0) return 0;  // sanity check
+    size_t x = *startfrom / 64;
+    if (x >= bitset->arraysize) {
+        return 0;  // nothing more to iterate over
+    }
+    uint64_t w = bitset->array[x];
+    w >>= (*startfrom & 63);
+    size_t howmany = 0;
+    size_t base = x << 6;
+    while (howmany < capacity) {
+        while (w != 0) {
+            uint64_t t = w & (~w + 1);
+            int r = roaring_trailing_zeroes(w);
+            buffer[howmany++] = r + base;
+            if (howmany == capacity) goto end;
+            w ^= t;
+        }
+        x += 1;
+        if (x == bitset->arraysize) {
+            break;
+        }
+        base += 64;
+        w = bitset->array[x];
+    }
+end:
+    if (howmany > 0) {
+        *startfrom = buffer[howmany - 1];
+    }
+    return howmany;
+}
 
+// return true if uninterrupted
+bool bitset_for_each(const bitset_t *b, bitset_iterator iterator,
```

Comment:
We want `bitset_for_each` to be inline.

Consider this code for example...

```C++
#include <cassert>
#include <cstdint>
#include <cstddef>

static inline int roaring_trailing_zeroes(unsigned long long input_num) { return __builtin_ctzll(input_num); }
struct bitset_s {
    uint64_t * array;
    /* For simplicity and performance, we prefer to have a size and a capacity that is a multiple of 64 bits.
     * Thus we only track the size and the capacity in terms of 64-bit words allocated */
    size_t arraysize;
    size_t capacity;
};

typedef struct bitset_s bitset_t;
typedef bool (*bitset_iterator)(size_t value, void *param);

 // return true if uninterrupted
 inline bool bitset_for_each(const bitset_t *b, bitset_iterator iterator,
                             void *ptr) {
     size_t base = 0;
     for (size_t i = 0; i < b->arraysize; ++i) {
         uint64_t w = b->array[i];
         while (w != 0) {
             uint64_t t = w & (~w + 1);
             int r = roaring_trailing_zeroes(w);
             if (!iterator(r + base, ptr)) return false;
             w ^= t;
         }
         base += 64;
     }
     return true;
 }

 bool sum(size_t value, void *param) {
    *(size_t*)param = value;
    return true;
 }

 size_t summe(const bitset_t *b) {
    size_t s = 0;
    bitset_for_each(b, sum, &s);
    return s;
 }
```

If you compile it with optimization, you will find that the bitset_iterator function gets inlined. That's only possible if the definition of `bitset_for_each` is available when you compile `summe`. Though it is possible to achieve the same result without inlining, it is much less clear how it would happen in general.